src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1487,52 +1487,99 @@
+                               (if-let [v (.get roots->facts roots-group)]
+                                 (.add ^java.util.List v fact)
+                                 (.put roots->facts roots-group (doto (java.util.LinkedList.)
+                                                                  (.add fact)))))]
 
     (fn [facts]
-      (for [[fact-type facts] (platform/group-by-seq fact-grouping-fn facts)]
-
-        (if-let [alpha-nodes (get @alpha-map fact-type)]
+      (let [roots->facts (java.util.LinkedHashMap.)]
 
-          ;; If the matching alpha nodes are cached, simply return them.
-          [alpha-nodes facts]
-
-          ;; The alpha nodes weren't cached for the type, so get them now.
-          (let [ancestors (conj (ancestors-fn fact-type) fact-type)
-
-                ;; Get all alpha nodes for all ancestors.  Keep them sorted to maintain
-                ;; deterministic ordering of fact propagation across the network.
-                ;; Alpha nodes do not have a :node-id of their own right now, so sort
-                ;; by the :node-id of their :children.
-                new-nodes (sort-by #(mapv :node-id (:children %))
-                                   (into []
-                                         (comp (map #(get alpha-roots %))
-                                               cat
-                                               (distinct))
-                                         ancestors))]
+        (doseq [fact facts
+                roots-group (fact-type->roots (wrapped-fact-type-fn fact))]
+          (update-roots->facts! roots->facts roots-group fact))
+        
+        (let [return-list (java.util.LinkedList.)
+              entries (.entrySet roots->facts)
+              entries-it (.iterator entries)]
+          ;; We iterate over the LinkedHashMap manually to avoid potential issues described at http://dev.clojure.org/jira/browse/CLJ-1738
+          ;; where a Java iterator can return the same entry object repeatedly and mutate it after each next() call.  We use mutable lists
+          ;; for performance but wrap them in unmodifiableList to make it clear that the caller is not expected to mutate these lists.
+          ;; Since after this function returns the only reference to the fact lists will be through the unmodifiedList we can depend elsewhere
+          ;; on these lists not changing.  Since the only expected workflow with these lists is to loop through them, not add or remove elements,
+          ;; we don't gain much from using a transient (which can be efficiently converted to a persistent data structure) rather than a mutable type. 
+          (loop []
+            (when (.hasNext entries-it)
+              (let [^java.util.Map$Entry e (.next entries-it)]
+                (.add return-list [(-> e ^AlphaRootsWrapper (.getKey) .roots)
+                                   (java.util.Collections/unmodifiableList (.getValue e))])
```

Comment:
I don't believe this is a reflection warning right?  I think the compiler can resolve the method since it only has one signature of this length.  Just double checking.